canton/community/ledger/ledger-api-core/src/main/scala/com/digitalasset/canton/platform/apiserver/services/admin/ApiPackageManagementService.scala

```diff
@@ -131,57 +133,138 @@ private[apiserver] final class ApiPackageManagementService private (
     } yield dar
   }
 
-  private def validateUpgrade(upgradingPackage: Dar[(Ref.PackageId, Ast.Package)])(implicit
-      loggingContext: LoggingContextWithTrace
+  private def lookupDar(pkgId: Ref.PackageId)(implicit
+                                                  loggingContext: LoggingContextWithTrace
+  ): Future[Option[(Ref.PackageId, Ast.Package)]] = {
+    for {
+      optArchive
+        <- packagesIndex.getLfArchive(pkgId)
+      optPackage
+        <- Future.fromTry {
+        optArchive.traverse(Decode.decodeArchive(_))
+          .handleError(Validation.handleLfArchiveError)
+      }
+    } yield optPackage
+  }
+
+  private def existingVersionedPackageId(dar: (Ref.PackageId, Ast.Package)): Option[Ref.PackageId] = {
+    val (_, pkg) = dar
+    val pkgName = pkg.metadata.name
+    val pkgVersion = pkg.metadata.version
+    packageMetadataStore.getSnapshot.getUpgradablePackageMap.collectFirst { case (pkgId, (`pkgName`, `pkgVersion`)) => pkgId }
+  }
+
+  private def minimalVersionedDar(upgradedPackageId: Ref.PackageId, dar: (Ref.PackageId, Ast.Package))(implicit
+                                                                loggingContext: LoggingContextWithTrace
+  ): Future[Option[(Ref.PackageId, Ast.Package)]] = {
+    val (_, pkg) = dar
+    val pkgName = pkg.metadata.name
+    val pkgVersion = pkg.metadata.version
+    packageMetadataStore.getSnapshot.getUpgradablePackageMap.collect { case (pkgId, (`pkgName`, pkgVersion)) =>
+      (pkgId, pkgVersion)
+    }
+      .filter { case (pkgId, version) => pkgVersion < version && upgradedPackageId != pkgId }
```

Comment:
Why do you need to check `upgradedPackageId != pkgId`? Shouldn't `upgradedPackageId == pkgId` hold only when `pkgVersion == version`?

Reply:
For `minimalVersionedDar`, we can leave this additional filter out. I'd originally added it to avoid double type checking. For example, during the following sequence:
- upload p1@v1, with no upgrade claim
- upload p2@v2, with claim to upgrade p1

In this latter case, we perform the type check that p2@v2 upgrades p1@v1 twice:
- once as part of Dylan's dar file based type checks
- and once as part of minimal dar file type checks.

If we drop this filter, then we still perform some unnecessary type checks (which I hadn't originally thought about).

I'll drop this additional check here as it's a premature optimisation.